File: Data-Structures/MaxHeap.js
/*
- Heap -

- It is complete: it cannot be unbalanced. We insert every new item to the next available place
- The heap is one maximally efficient implementation of a priority queue abstract data type
- It has nothing to do with the pool of memory from which dynamically allocated memory is allocated

- Basically a binary tree
- Either min heap or max heap
- Constructed from left to right

Min Heap:
- Parent nodes are always less then or equal to child nodes
- Lowest node is the root
Max Heap:
- Parent nodes are always more than or equal to child nodes
- Highest node is the root

Applications:

- Dijkstra's algorithm, Prims algorithm
*/

class MaxHeap {
    constructor(capacity) {
        this.capacity = capacity
        this.heap = Array(capacity).fill(null)
        this.size = 0
    }
    insert(item) {
        if (this.capacity === this.size){
            return
        }
        this.heap[this.size] = item
        this.size++
        this.fixUp(this.size - 1)
    }
    fixUp(index) {
        let pIndex = Math.floor((index - 1) / 2)
        if (index > 0 && this.heap[pIndex] < this.heap[index]) {
            this.swap(pIndex, index)
            this.fixUp(pIndex)
        }
    }
    swap(i, j) {
        let swap = this.heap[i]
        this.heap[i] = this.heap[j]
        this.heap[j] = swap
    }
    poll() {
        const data = this.heap[0]
        this.swap(0, this.size - 1)
        this.size--
        this.fixDown(0)
        return data
    }
    fixDown(index) {
        let leftIndex = index * 2 + 1
        let rightIndex = index * 2 + 2
        let maxIndex = index
        if (leftIndex < this.size && this.heap[leftIndex] > this.heap[maxIndex]) {
            maxIndex = leftIndex
        }
        if (rightIndex < this.size && this.heap[rightIndex] > this.heap[maxIndex]) {
            maxIndex = rightIndex
        }
        if (maxIndex !== index) {
            this.swap(index, maxIndex)
            this.fixDown(maxIndex)
        }
    }
    heapSort() {
        const range = this.size
        for (i = 0; i < range; i++) {
            this.poll()
        }
    }
}

const heap = new MaxHeap(10)
const arr = [10, 8, 12, 20, -2, 0, 1, 321, 44, 55]
for (i = 0; i < arr.length; i++) {
    heap.insert(arr[i])
}

heap.heapSort()
console.log(heap.heap)
console.log(heap.size)

File: DataStructures/AssociativeArrays/MaxHeap.py
"""
- Heap -

- It is complete: it cannot be unbalanced. We insert every new item to the next available place
- The heap is one maximally efficient implementation of a priority queue abstract data type
- It has nothing to do with the pool of memory from which dynamically allocated memory is allocated

- Basically a binary tree
- Either min heap or max heap
- Constructed from left to right

Min Heap:
- Parent nodes are always less then or equal to child nodes
- Lowest node is the root
Max Heap:
- Parent nodes are always more than or equal to child nodes
- Highest node is the root

Applications:

- Dijkstra's algorithm, Prims algorithm
"""

class MaxHeap:
    def __init__(self, capacity):
        self.capacity = capacity
        self.heap = [None] * capacity
        self.size = 0

    def swap(self, i, j):
        self.heap[i], self.heap[j] = self.heap[j], self.heap[i]

    def insert(self, item):
        if self.size == self.capacity:
            return
        self.heap[self.size] = item
        self.size += 1
        self.fix_up(self.size - 1)

    def fix_up(self, index):
        p_index = (index - 1) // 2
        if index > 0 and self.heap[index] > self.heap[p_index]:
            self.swap(index, p_index)
            self.fix_up(p_index)

    def poll(self):
        data = self.heap[0]
        self.swap(0, self.size - 1)
        self.size -= 1
        self.fix_down(0)
        return data

    def fix_down(self, index):
        l_index = index * 2 + 1
        r_index = index * 2 + 2
        max_index = index
        if l_index < self.size and self.heap[l_index] > self.heap[max_index]:
            max_index = l_index
        if r_index < self.size and self.heap[r_index] > self.heap[max_index]:
            max_index = r_index
        if max_index != index:
            self.swap(index, max_index)
            self.fix_down(max_index)

    def heap_sort(self):
        for i in range(self.size):
            self.poll()


CAPACITY = 10
heap = MaxHeap(CAPACITY)
array = [10, 8, 12, 20, -2, 0, 1, 321, 44, 55]
for num in array:
    heap.insert(num)

heap.heap_sort()
print('after sort')
print(heap.heap)
print(heap.size)
